pkg/devfile/parser/resolutionContext.go

```diff
@@ -0,0 +1,67 @@
+package parser
+
+import (
+	"fmt"
+	"reflect"
+
+	v1 "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+)
+
+// resolutionContextTree is a recursive structure representing information about the devfile that is
+// lost when flattening (e.g. plugins, parents)
+type resolutionContextTree struct {
+	importReference v1.ImportReference
+	parentNode      *resolutionContextTree
+}
+
+// appendNode adds a new node to the resolution context.
+func (t *resolutionContextTree) appendNode(importReference v1.ImportReference) *resolutionContextTree {
+	newNode := &resolutionContextTree{
+		importReference: importReference,
+		parentNode:      t,
+	}
+	return newNode
+}
+
+// hasCycle checks if the current resolutionContextTree has a cycle
+func (t *resolutionContextTree) hasCycle() error {
+	var seenRefs []v1.ImportReference
+	currNode := t
+	for currNode.parentNode != nil {
+		for _, seenRef := range seenRefs {
+			if reflect.DeepEqual(seenRef, currNode.importReference) {
+				return fmt.Errorf("devfile has an cycle in references: %v", formatImportCycle(t))
```

Comment:
we do print out the cycle. check function `formatImportCycle()`

Reply:
the loop in `hasCycle()` and `formatImportCycle()` are different. `hasCycle()` returns once it found the current importReference has dup. but `formatImportCycle()` want to traverse to the real end, which is the main devfile. 

for example, for an import cycle: `main devfile -> url1 -> url2 -> url3 -> url2`
`formatImportCycle()` will printout the entire import cycle
but `hascyle()` will only tracks `url2->url3->url2`, and lost track of the previous nodes. (because it traverse from the end)

Reply:
I meant like this where its only one list traversal, which also works. But I will leave it up to you how you prefer it.

```go
// hasCycle checks if the current resolutionContextTree has a cycle
func (t *resolutionContextTree) hasCycle() error {
	var seenRefs []v1.ImportReference
	currNode := t
	hasCycle := false
	cycle := resolveImportReference(t.importReference)

	for currNode.parentNode != nil {
		for _, seenRef := range seenRefs {
			if reflect.DeepEqual(seenRef, currNode.importReference) {
				hasCycle = true
			}
		}
		seenRefs = append(seenRefs, currNode.importReference)
		currNode = currNode.parentNode
		cycle = fmt.Sprintf("%s -> %s", resolveImportReference(currNode.importReference), cycle)
	}

	if hasCycle {
		return fmt.Errorf("devfile has an cycle in references: %v", cycle)
	}
	return nil
}
```

Reply:
right, this also works. updated